### PR TITLE
feat: hasac

### PR DESCRIPTION
**Pull request by sash-a**
## What?
Heterogeneous agent soft actor critic: https://arxiv.org/pdf/2306.10715

Also moved the `step` function is all the SAC methods to a more sensible place

### CODE REVIEW

**Comment by sash-a** [comment]
Easier to review the diff between masac and hasac: https://www.diffchecker.com/EKMo2FcC/

File: mava/systems/sac/anakin/ff_hasac.py [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+    actor_network = Actor(actor_torso, actor_action_head)
+    actor_params = jax.vmap(actor_network.init, in_axes=(0, None))(actor_keys, obs_single_batched)
+
+    # Making Q networks
+    critic_torso = hydra.utils.instantiate(cfg.network.critic_network.pre_torso)
+    q_network = QNetwork(critic_torso, centralised_critic=True)
+    q1_params = q_network.init(q1_key, obs_single_batched, concat_acts_batched)
+    q2_params = q_network.init(q2_key, obs_single_batched, concat_acts_batched)
+    q1_target_params = q_network.init(q1_target_key, obs_single_batched, concat_acts_batched)
+    q2_target_params = q_network.init(q2_target_key, obs_single_batched, concat_acts_batched)
```
Comment (RuanJohn) [inline mava/systems/sac/anakin/ff_hasac.py]: Just adding a note here to double check whether the critics should condition on a single agent's observation and the joint observation along with all the actions.

File: mava/systems/sac/anakin/ff_hasac.py [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+from mava.wrappers import episode_metrics
+
+
+# It is faster to do this with a vmap, but unfortunately that requires switching to numpyro.
+# This requires a lot of testing so there is currently an issue for it: #1098
+def get_actions(
+    actor_params: FrozenVariableDict,
+    actor_net: Actor,
+    keys: chex.PRNGKey,
+    env: MarlEnv,
```
Comment (RuanJohn) [inline mava/systems/sac/anakin/ff_hasac.py]: Passing in env just to get the number of agents seems a bit much. Can't we just set num_agents in the config and use it from there or something along those lines.

File: mava/systems/sac/anakin/ff_hasac.py [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+        params: SacParams, opt_states: OptStates, data: Transition, key: chex.PRNGKey
+    ) -> Tuple[SacParams, OptStates, Metrics]:
+        """Update the Q parameters."""
+        # Calculate Q target values.
+        act_keys = jax.random.split(key, env.num_agents)
+        next_action, next_log_prob = get_actions(
+            params.actor, actor_net, act_keys, env, data.next_obs
+        )
+
+        joint_next_actions = get_joint_action(next_action)
```
Comment (RuanJohn) [inline mava/systems/sac/anakin/ff_hasac.py]: A general note, but I think adding shapes in the comments could be super useful here.

File: mava/systems/sac/anakin/ff_hasac.py [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+    def update_actor_and_alpha(
+        params: SacParams, opt_states: OptStates, data: Transition, key: chex.PRNGKey
+    ) -> Tuple[SacParams, OptStates, Metrics]:
+        """Update the actor and alpha parameters. Compensated for the delay in policy updates."""
+        alpha_grad_fn = jax.value_and_grad(alpha_loss_fn)
+        actor_grad_fn = jax.value_and_grad(actor_loss_fn)
+
+        # compensate for the delay by doing `policy_frequency` updates instead of 1.
+        assert cfg.system.policy_update_delay > 0, "Need to have a policy update delay > 0."
+        for _ in range(cfg.system.policy_update_delay):
```
Comment (RuanJohn) [inline mava/systems/sac/anakin/ff_hasac.py:390]: Is it standard to do multiple policy updates after delaying like this?

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Overall looking very nice! 
Something that would be great is to add some comments on the shapes. And then just a questions and suggestions here and there.

File: mava/systems/sac/anakin/ff_hasac.py (hunk 4 above)
Comment (sash-a) [reply mava/systems/sac/anakin/ff_hasac.py:390]: Ye it's a TD3 thing that people found useful in SAC, CleanRL does it [here](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/sac_continuous_action.py#L265)

File: mava/systems/sac/anakin/ff_hasac.py (hunk 1 above)
Comment (sash-a) [reply mava/systems/sac/anakin/ff_hasac.py]: So I just had a look and this actually will give us the state. The env is created with `global_state=True` always and so the critic will always pull out the global in favour of the obs.

Sorry for the confusion yesterday

Comment (RuanJohn) [reply mava/systems/sac/anakin/ff_hasac.py]: Ah okay, makes sense. Thanks for clarifying :raised_hands: 

Can you just add a comment here, that the obs object contains the global state please?

File: mava/systems/sac/anakin/ff_hasac.py (hunk 4 above)
Comment (RuanJohn) [reply mava/systems/sac/anakin/ff_hasac.py:390]: Perfect, thanks for clarifying!

File: mava/systems/sac/anakin/ff_hasac.py [hunk 5]
```diff
@@ -0,0 +1,717 @@
...
+            key, act_key, agent_order_key = jax.random.split(key, 3)
+            act_keys = jax.random.split(act_key, env.num_agents)
+            if cfg.system.shuffle_agents:
+                agent_ids = jax.random.permutation(agent_order_key, env.num_agents)
+            else:
+                agent_ids = jnp.arange(env.num_agents)
+
+            joint_actions, log_probs = get_actions(
```
Comment (RuanJohn) [inline mava/systems/sac/anakin/ff_hasac.py:400]: Shapes for joint_actions and log_probs here would be great.

File: mava/systems/sac/anakin/ff_hasac.py [hunk 6]
```diff
@@ -0,0 +1,717 @@
...
+            )
+
+            # HASAC sequential update: run the normal actor update one at a time instead of batched.
+            # Update the joint actions after updating the actor and use the new joint actions.
+            for agent_id in agent_ids:
+                key, actor_key = jax.random.split(key)
+
+                agent_params = tree_slice(params.actor, agent_id)
+                agent_opt_state = tree_slice(opt_states.actor, agent_id)
+                agent_obs = tree_slice(data.obs, jnp.s_[:, agent_id])
```
Comment (RuanJohn) [inline mava/systems/sac/anakin/ff_hasac.py:414]: Similarly, a shape here would be nice. Also what is `jnp.s_`? I haven't seen it before.

**Review by RuanJohn — COMMENTED** [review]
Just two other places where I think a shape would be nice. And then a question on `jnp.s_`

File: mava/systems/sac/anakin/ff_hasac.py (hunk 6 above)
Comment (sash-a) [reply mava/systems/sac/anakin/ff_hasac.py:414]: Good question, it returns what's called an index expression, which is basically a way to pass an index eg `[:, 1]` around as a variable. So here it's convenient to pass it to `tree_slice` instead of writing `tree.map(lambda x: x[:, agent_id], data.obs)`

File: mava/systems/sac/anakin/ff_hasac.py [hunk 7]
```diff
@@ -0,0 +1,727 @@
...
+
+    full_action_shape = (cfg.arch.num_envs, *env.action_spec().shape)
+
+    # losses:
+    def q_loss_fn(
+        q_params: QVals, obs: Array, action: Array, target: Array
+    ) -> Tuple[Array, Metrics]:
+        q1_params, q2_params = q_params
+        # Concat all actions and tile them for num agents to create joint actions for all agents
+        joint_action = get_joint_action(action)  # (B, A, Act) -> (B, A, A * Act)
```
Comment (RuanJohn) [inline mava/systems/sac/anakin/ff_hasac.py]: Just a note to please update all the shape comments. Can we please do this everywhere? 
```suggestion
        joint_action = get_joint_action(action)  # (B, N, A) -> (N, N, N * A)
```

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Last minor nits on the shape comments.

**Review by RuanJohn — DISMISSED** [review]
This is amazing. Great work @sash-a :fire:

File: mava/configs/arch/anakin.yaml [hunk 8]
```diff
@@ -1,13 +1,13 @@
 # --- Anakin config ---
 
 # --- Training ---
-num_envs: 16  # Number of vectorised environments per device.
+num_envs: 32  # Number of vectorised environments per device.
```
Comment (OmaymaMahjoub) [inline mava/configs/arch/anakin.yaml]: can you revert the changes here to keep it to the default 16 envs

File: mava/utils/jax_utils.py [hunk 9]
```diff
@@ -14,13 +14,26 @@
...
 import jax
 import jax.numpy as jnp
 import numpy as np
 from jax import tree
+from typing_extensions import TypeAlias
+
+Indexer: TypeAlias = Union[int, slice, Tuple[slice, ...]]
+
+
+def tree_slice(pytree: chex.ArrayTree, i: Indexer) -> chex.ArrayTree:
```
Comment (OmaymaMahjoub) [inline mava/utils/jax_utils.py:30]: can you add documentation here to explain the need for this fn

File: mava/utils/jax_utils.py [hunk 10]
```diff
@@ -14,13 +14,26 @@
...
+    return tree.map(lambda x: x[i], pytree)
+
+
+def tree_at_set(old_tree: chex.ArrayTree, i: Indexer, new_tree: chex.ArrayTree) -> chex.ArrayTree:
```
Comment (OmaymaMahjoub) [inline mava/utils/jax_utils.py:35]: same for this util function

File: mava/systems/sac/anakin/ff_hasac.py [hunk 11]
```diff
@@ -0,0 +1,728 @@
...
+def hydra_entry_point(cfg: DictConfig) -> float:
+    """Experiment entry point."""
+    # Allow dynamic attributes.
+    OmegaConf.set_struct(cfg, False)
+    cfg.logger.system_name = "ff_hasac"
+
+    # Run experiment.
+    final_return = run_experiment(cfg)
+
+    print(f"{Fore.CYAN}{Style.BRIGHT}MASAC experiment completed{Style.RESET_ALL}")
```
Comment (OmaymaMahjoub) [inline mava/systems/sac/anakin/ff_hasac.py]:
```suggestion
    print(f"{Fore.CYAN}{Style.BRIGHT}HASAC experiment completed{Style.RESET_ALL}")
```

File: mava/systems/sac/anakin/ff_hasac.py [hunk 12]
```diff
@@ -0,0 +1,728 @@
...
+    obs_single_batched = tree.map(lambda x: x[0][jnp.newaxis, ...], obs)
+
+    # Making actor network
+    actor_torso = hydra.utils.instantiate(cfg.network.actor_network.pre_torso)
+    action_head, _ = get_action_head(env)
+    actor_action_head = hydra.utils.instantiate(
+        action_head, action_dim=env.action_dim, independent_std=False
+    )
+    actor_network = Actor(actor_torso, actor_action_head)
+    actor_params = jax.vmap(actor_network.init, in_axes=(0, None))(actor_keys, obs_single_batched)
```
Comment (OmaymaMahjoub) [inline mava/systems/sac/anakin/ff_hasac.py:162]: if you can add comment here to explain that the vmap used to produce different params per agent

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Looks good from my side, just it lacks some  documentations here and there to make it more clear and easy for users to follow the difference between hasac and the other sac systems

File: mava/systems/sac/anakin/ff_hasac.py [hunk 13]
```diff
@@ -0,0 +1,728 @@
...
+    key, eval_key = jax.random.split(key)
+
+    def eval_act_fn(
+        params: FrozenDict, timestep: TimeStep, key: chex.PRNGKey, actor_state: ActorState
+    ) -> Tuple[Action, Dict]:
+        keys = jax.random.split(key, eval_env.num_agents)
+        action, _ = get_actions(
+            params, actor, keys, eval_env.num_agents, eval_env.action_dim, timestep.observation
+        )
+        return action, {}
```
Comment (OmaymaMahjoub) [inline mava/systems/sac/anakin/ff_hasac.py:625]: why don't we do similar to ppo systems (not sure if it is the same for other sac systems), we import make_act_sac_fn from the evaluator and here we give it get actions as input

Comment (sash-a) [reply mava/systems/sac/anakin/ff_hasac.py:625]: Would be nice but it's not possible here because `get_actions` returns the actions and that function expects distributions. It would be possible if we had vmappable distributions, but I think that is for another PR.

**Review by RuanJohn — APPROVED** [review]
:fire:

**Review by OmaymaMahjoub — APPROVED** [review]
🚀

File: mava/systems/sac/anakin/ff_hasac.py (hunk 13 above)
Comment (OmaymaMahjoub) [reply mava/systems/sac/anakin/ff_hasac.py:625]: we decided to not do this changes, since the eval_act_fn only used for HASAC so keeping it on the system side is fine
